### ISSUE TITLE
Write test cases

### ISSUE THREAD

**Issue by mattiaisgro** [chaotic-society/theoretica#53]
Write test cases for all major blocks of the library using the Chebyshev testing framework. The functionalities which need test cases are:
- [ ] core
- [ ] algebra (more tests for transformations may be needed)
- [ ] autodiff
- [ ] polynomial 
- [ ] calculus
- [ ] complex
- [ ] interpolation
- [ ] optimization
- [ ] pseudorandom
- [ ] statistics

Benchmarks for performance critical code are also needed.

**Comment by Fahad-Ali-Khan-ca** [comment]
Hi, I would like to work on the test cases.

**Comment by mattiaisgro** [comment]
> Hi, I would like to work on the test cases.

Hello Fahad, thank you for your interest in the project!
Testing is a fundamental part of developing reliable software, so your help is much appreciated.

Theoretica is tested using [Chebyshev](https://github.com/chaotic-society/chebyshev), a testing framework that we developed specifically for this purpose. I suggest that you start by looking at Chebyshev and how it works, you may look at [Examples](https://github.com/chaotic-society/chebyshev/tree/main/examples) and [Documentation](chaotic-society.github.io/chebyshev).

**I'm going to update the existing code and issue description to make it easier for you to participate. After that, you may fork the project and start writing test cases!** If you like the project, please leave a star. Feel free to ask me any questions you have!

**Comment by Fahad-Ali-Khan-ca** [comment]
Hi, thanks for all the help making it easier for me. Just one quick question. Earlier the issue had check marked some modules like core and autodiff. Do you need me to make test cases for them as well?

**Comment by mattiaisgro** [comment]
> Hi, thanks for all the help making it easier for me. Just one quick question. Earlier the issue had check marked some modules like core and autodiff. Do you need me to make test cases for them as well?

Yes, some modules already have test cases written, such as the `core` module, but the tests are not complete, some functionalities are not checked. So I have updated the checklist to include new modules, we will check them when most of the testable functionalities are complete. Tests for `signal` are almost complete, as there are only two functions right now.

In the last commit I have added files named `test_MODULE.cpp` for each module, so that you have a starting point and the build system is already set up. When you feel confident, you can fork the repo and start writing test cases. I would start by writing tests for the `core`, `calculus` and `complex` modules as they are easier to test.

**Comment by Fahad-Ali-Khan-ca** [comment]
Thanks for the support I have forked the repo!

**Comment by mattiaisgro** [comment]
I have just added test cases for the `signal` module and they are now complete! Marked off the list.

**Comment by mattiaisgro** [comment]
With the new tests in `test_calculus.cpp`, the `calculus` module is also complete.

**Comment by mattiaisgro** [comment]
Tests for the `complex` module are now complete.

**Comment by mattiaisgro** [comment]
Tests for the `io` module are already implemented in the new merge commit.

**Comment by mattiaisgro** [comment]
Tests for the `polynomial` module have been added in https://github.com/chaotic-society/theoretica/commit/48dc4242041d66671983a2fb0761a89ac176602d.

**Comment by mattiaisgro** [comment]
Tests for the `autodiff` module have been added in https://github.com/chaotic-society/theoretica/commit/d94d6a9ef26d873cb71ca142112b733643eb79eb.

**Comment by mattiaisgro** [comment]
Tests for the `optimization` module have been added in https://github.com/chaotic-society/theoretica/commit/57bb32c51a9be8b404d77aadc6f6704429b5324a.

**Comment by mattiaisgro** [comment]
Tests for the `algebra` module have been added in https://github.com/chaotic-society/theoretica/commit/6392c438fdc983fc3e74e1b15fe294126e3ab2ad.
